File: server/src/collectors/jmaUv.js
/**
 * JMA UV index collector (jma-uv) — kind:'intel'.
 *
 * JMA publishes the UV index ("紫外線情報") only as rendered raster maps /
 * an interactive viewer; there is no key-free machine-readable JSON endpoint
 * (every documented/guessed data path — data.jma.go.jp/env/uvhp/data/*.json,
 * bosai/uv/data/* — returns 404 as of this writing). Rather than fabricate
 * point values we emit a single portal intel item that records the official
 * UV-index product URL and a best-effort reachability probe.
 *
 * NOTE: the JSON endpoint is UNVERIFIED / believed not to exist publicly; if
 * JMA later exposes one this collector should be upgraded to parse it.
 */

import { fetchHead } from './_liveHelpers.js';
import { intelEnvelope, intelUid } from '../utils/intelHelpers.js';

const SOURCE_ID = 'jma-uv';
const PORTAL = 'https://www.jma.go.jp/jp/uv/';

export default async function collectJmaUv() {
  const live = await fetchHead(PORTAL, { timeoutMs: 8000 }).catch(() => false);
  return intelEnvelope({
    sourceId: SOURCE_ID,
    items: [{
      uid: intelUid(SOURCE_ID, 'portal'),
      title: 'JMA UV index (紫外線情報)',
      summary: 'Nationwide ultraviolet index forecast and analysis maps (JMA)',
      body: 'JMA publishes the UV index only as raster maps / an interactive viewer; no key-free machine-readable JSON endpoint is available, so no point features are emitted.',
      link: PORTAL,
      author: '気象庁 Japan Meteorological Agency',
      language: 'ja',
      published_at: new Date().toISOString(),
      tags: ['uv', 'jma', 'health', live ? 'reachable' : 'unreachable'],
      properties: { operator: '気象庁', reachable: !!live, machine_readable: false },
    }],
    live: !!live,
    description: 'JMA UV index portal (no public JSON endpoint — portal item only)',
  });
}

File: native/collectors/sources/jma_uv.c
/* collectors/environment/sources/jma_uv.c
 * INTEL/probe source — port of server/src/collectors/jmaUv.js.
 * JMA UV index has no key-free JSON endpoint; emit ONE portal intel item
 * with a reachability probe (uid = jma-uv|portal). Honest — never fabricated. */
#include "../../source.h"
#include "../../lib/probe.h"
#include "../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>

#define PORTAL "https://www.jma.go.jp/jp/uv/"

static int run(const source_ctx *ctx, intel_sink *sink) {
  int live = probe_head(ctx->http, PORTAL);
  char now[32]; probe_iso_now(now, sizeof now);

  cJSON *tags = cJSON_CreateArray();
  cJSON_AddItemToArray(tags, cJSON_CreateString("uv"));
  cJSON_AddItemToArray(tags, cJSON_CreateString("jma"));
  cJSON_AddItemToArray(tags, cJSON_CreateString("health"));
  cJSON_AddItemToArray(tags, cJSON_CreateString(live ? "reachable" : "unreachable"));
  char *tj = cJSON_PrintUnformatted(tags);

  cJSON *p = cJSON_CreateObject();
  cJSON_AddStringToObject(p, "operator", "\xE6\xB0\x97\xE8\xB1\xA1\xE5\xBA\x81");
  cJSON_AddBoolToObject(p, "reachable", live);
  cJSON_AddBoolToObject(p, "machine_readable", 0);
  char *pj = cJSON_PrintUnformatted(p);

  intel_item it = {0};
  it.remote_key = "portal";
  it.title = "JMA UV index (\xE7\xB4\xAB\xE5\xA4\x96\xE7\xB7\x9A\xE6\x83\x85\xE5\xA0\xB1)";
  it.summary = "Nationwide ultraviolet index forecast and analysis maps (JMA)";
  it.body = "JMA publishes the UV index only as raster maps / an interactive "
            "viewer; no key-free machine-readable JSON endpoint is available, "
            "so no point features are emitted.";
  it.link = PORTAL;
  it.author = "\xE6\xB0\x97\xE8\xB1\xA1\xE5\xBA\x81 Japan Meteorological Agency";
  it.lang = "ja";
  it.published_at = now;
  it.tags_json = tj;
  it.properties_json = pj;
  int rc = sink->emit(sink, &it);

  free(tj); free(pj);
  cJSON_Delete(tags); cJSON_Delete(p);
  fprintf(stderr, "[jma-uv] probe reachable=%d\n", live);
  return rc >= 0 ? 0 : -1;
}

static const source_def jma_uv_def = {
  .id = "jma-uv", .collector = "environment",
  .name = "JMA UV Index", .name_ja = "気象庁 紫外線情報",
  .update_interval_sec = 3600, .run = run,
};
REGISTER_SOURCE(jma_uv_def)
